File: packages/backend/src/core/tokenRefresh.ts
/**
 * Token 刷新模块
 * 从 Electron 主进程提取的 Token 刷新逻辑
 */

import { createLogger } from '../utils/logger.js'
import type { ProxyAccount, TokenRefreshResult } from './types.js'
import { getKiroUserAgent } from './kiroApi.js'

const logger = createLogger('TokenRefresh')

// Kiro Auth 端点
const KIRO_AUTH_ENDPOINT = 'https://prod.us-east-1.auth.desktop.kiro.dev'

/**
 * OIDC Token 刷新 (BuilderId/IdC)
 * 用于 AWS SSO 认证方式
 */
export async function refreshOidcToken(
  refreshToken: string,
  clientId: string,
  clientSecret: string,
  region: string = 'us-east-1'
): Promise<TokenRefreshResult> {
  logger.info(`Refreshing OIDC token, region: ${region}`)

  const url = `https://oidc.${region}.amazonaws.com/token`

   const payload = {
    clientId,
    clientSecret,
    refreshToken,
    grantType: 'refresh_token'
  }

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    })

    if (!response.ok) {
      const errorText = await response.text()
      logger.error(`OIDC refresh failed: ${response.status} - ${errorText}`)
      return { success: false, error: `HTTP ${response.status}: ${errorText}` }
    }

    const data = await response.json() as {
      accessToken: string
      refreshToken?: string
      expiresIn?: number
    }

    logger.info(`OIDC token refreshed successfully, expires in ${data.expiresIn}s`)

    return {
      success: true,
      accessToken: data.accessToken,
      refreshToken: data.refreshToken || refreshToken,
      expiresAt: Date.now() + (data.expiresIn || 3600) * 1000
    }
  } catch (error) {
    logger.error(`OIDC refresh error: ${error}`)
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    }
  }
}

/**
 * 社交登录 Token 刷新 (GitHub/Google)
 * 使用 Kiro Auth Service
 * @param refreshToken - 刷新令牌
 * @param machineId - 机器码（可选，用于生成 User-Agent）
 */
export async function refreshSocialToken(
  refreshToken: string,
  machineId?: string
): Promise<TokenRefreshResult> {
  logger.info('Refreshing social token')

  const url = `${KIRO_AUTH_ENDPOINT}/refreshToken`

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        // 关键修复：使用包含 machineId 的 User-Agent（与 kiro-m 保持一致）
        'User-Agent': getKiroUserAgent(machineId)
      },
      body: JSON.stringify({ refreshToken })
    })

    if (!response.ok) {
      const errorText = await response.text()
      logger.error(`Social refresh failed: ${response.status} - ${errorText}`)
      return { success: false, error: `HTTP ${response.status}: ${errorText}` }
    }

    const data = await response.json() as {
      accessToken: string
      refreshToken?: string
      expiresIn?: number
    }

    logger.info(`Social token refreshed successfully, expires in ${data.expiresIn}s`)

    return {
      success: true,
      accessToken: data.accessToken,
      refreshToken: data.refreshToken || refreshToken,
      expiresAt: Date.now() + (data.expiresIn || 3600) * 1000
    }
  } catch (error) {
    logger.error(`Social refresh error: ${error}`)
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    }
  }
}

/**
 * 根据认证方式刷新 Token
 */
export async function refreshTokenByMethod(
  account: ProxyAccount
): Promise<TokenRefreshResult> {
  if (!account.refreshToken) {
    return { success: false, error: 'No refresh token available' }
  }

  // 社交登录使用 Kiro Auth Service
  if (account.authMethod === 'social') {
    // 关键修复：传递 machineId 给 refreshSocialToken（与 kiro-m 保持一致）
    return refreshSocialToken(account.refreshToken, account.machineId)
  }

  // IdC/BuilderId 使用 OIDC
  if (!account.clientId || !account.clientSecret) {
    return { success: false, error: 'Missing clientId or clientSecret for OIDC refresh' }
  }

  return refreshOidcToken(
    account.refreshToken,
    account.clientId,
    account.clientSecret,
    account.region || 'us-east-1'
  )
}

/**
 * 检查 Token 是否需要刷新
 * @param account 账号
 * @param beforeExpirySec 提前刷新秒数（默认5分钟）
 */
export function needsTokenRefresh(
  account: ProxyAccount,
  beforeExpirySec: number = 300
): boolean {
  if (!account.expiresAt) {
    return false // 没有过期时间，不刷新
  }

  const now = Date.now()
  const refreshThreshold = account.expiresAt - beforeExpirySec * 1000

  return now >= refreshThreshold
}

/**
 * 检查 Token 是否已过期
 */
export function isTokenExpired(account: ProxyAccount): boolean {
  if (!account.expiresAt) {
    return false // 没有过期时间，假设未过期
  }

  return Date.now() >= account.expiresAt
}

File: packages/gateway/core/token_refresh.go
package core

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/rs/zerolog/log"

	"github.com/kiro-gateway/gateway/config"
	"github.com/kiro-gateway/gateway/model"
	"github.com/kiro-gateway/gateway/util"
)

const (
	// KiroAuthServiceURL is the base URL for the Kiro social authentication service.
	KiroAuthServiceURL = "https://prod.us-east-1.auth.desktop.kiro.dev"

	// OIDCTokenPath is the path for OIDC token endpoint.
	OIDCTokenPath = "/token"

	// tokenRefreshTimeout is the HTTP timeout for token refresh requests.
	tokenRefreshTimeout = 30 * time.Second
)

// socialRefreshResponse is the JSON response from the social token refresh endpoint.
type socialRefreshResponse struct {
	AccessToken  string `json:"accessToken"`
	RefreshToken string `json:"refreshToken"`
	ExpiresAt    int64  `json:"expiresAt"`
}

// oidcTokenResponse is the JSON response from the OIDC token endpoint.
type oidcTokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresIn    int64  `json:"expires_in"`
	TokenType    string `json:"token_type"`
}

// RefreshTokenByMethod dispatches to the appropriate token refresh handler
// based on the account's authentication method.
func RefreshTokenByMethod(ctx context.Context, account *model.ProxyAccount) (*model.TokenRefreshResult, error) {
	if account.RefreshToken == "" {
		return &model.TokenRefreshResult{
			Success: false,
			Error:   "no refresh token available",
		}, fmt.Errorf("no refresh token for account %s", account.ID)
	}

	switch account.AuthMethod {
	case "idc", "builderId":
		return refreshOIDCToken(ctx, account)
	default:
		// Default to social auth (includes "social" and empty authMethod).
		return refreshSocialToken(ctx, account)
	}
}

// refreshSocialToken refreshes an access token using the Kiro social auth endpoint.
func refreshSocialToken(ctx context.Context, account *model.ProxyAccount) (*model.TokenRefreshResult, error) {
	reqURL := KiroAuthServiceURL + "/refresh"

	body := fmt.Sprintf(`{"refreshToken":"%s"}`, account.RefreshToken)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, reqURL, strings.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create social refresh request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", fmt.Sprintf("kiro-gateway/%s", account.MachineID))

	client := util.NewHTTPClient(account.ProxyURL, tokenRefreshTimeout)
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("social refresh request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read social refresh response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		log.Warn().
			Str("accountId", account.ID).
			Int("status", resp.StatusCode).
			Str("body", string(respBody)).
			Msg("Social token refresh failed")
		return &model.TokenRefreshResult{
			Success: false,
			Error:   fmt.Sprintf("HTTP %d: %s", resp.StatusCode, string(respBody)),
		}, fmt.Errorf("social refresh HTTP %d", resp.StatusCode)
	}

	var result socialRefreshResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("parse social refresh response: %w", err)
	}

	log.Info().
		Str("accountId", account.ID).
		Int64("expiresAt", result.ExpiresAt).
		Msg("Social token refreshed")

	return &model.TokenRefreshResult{
		Success:      true,
		AccessToken:  result.AccessToken,
		RefreshToken: result.RefreshToken,
		ExpiresAt:    result.ExpiresAt,
	}, nil
}

// refreshOIDCToken refreshes an access token using the OIDC token endpoint.
func refreshOIDCToken(ctx context.Context, account *model.ProxyAccount) (*model.TokenRefreshResult, error) {
	region := account.Region
	if region == "" {
		region = "us-east-1"
	}

	oidcURL := fmt.Sprintf("https://oidc.%s.amazonaws.com%s", region, OIDCTokenPath)

	// Build form data.
	formData := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {account.RefreshToken},
		"client_id":     {account.ClientID},
	}
	if account.ClientSecret != "" {
		formData.Set("client_secret", account.ClientSecret)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, oidcURL, strings.NewReader(formData.Encode()))
	if err != nil {
		return nil, fmt.Errorf("create OIDC refresh request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	client := util.NewHTTPClient(account.ProxyURL, tokenRefreshTimeout)
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("OIDC refresh request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read OIDC refresh response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		log.Warn().
			Str("accountId", account.ID).
			Int("status", resp.StatusCode).
			Str("body", string(respBody)).
			Msg("OIDC token refresh failed")
		return &model.TokenRefreshResult{
			Success: false,
			Error:   fmt.Sprintf("HTTP %d: %s", resp.StatusCode, string(respBody)),
		}, fmt.Errorf("OIDC refresh HTTP %d", resp.StatusCode)
	}

	var result oidcTokenResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("parse OIDC refresh response: %w", err)
	}

	// Calculate absolute expiry time from relative expires_in.
	expiresAt := time.Now().UnixMilli() + (result.ExpiresIn * 1000)

	log.Info().
		Str("accountId", account.ID).
		Int64("expiresIn", result.ExpiresIn).
		Int64("expiresAt", expiresAt).
		Msg("OIDC token refreshed")

	return &model.TokenRefreshResult{
		Success:      true,
		AccessToken:  result.AccessToken,
		RefreshToken: result.RefreshToken,
		ExpiresAt:    expiresAt,
	}, nil
}

// NeedsTokenRefresh checks whether an account's token should be proactively refreshed
// before it expires, based on the configured threshold.
func NeedsTokenRefresh(account *model.ProxyAccount) bool {
	if account.ExpiresAt == 0 {
		return false
	}
	threshold := config.GetConfig().Proxy.TokenRefreshBeforeExpiry
	// ExpiresAt is in milliseconds; threshold is in seconds.
	return time.Now().UnixMilli()+int64(threshold)*1000 > account.ExpiresAt
}

// IsTokenExpired checks whether an account's token has already expired.
func IsTokenExpired(account *model.ProxyAccount) bool {
	if account.ExpiresAt == 0 {
		return false
	}
	return time.Now().UnixMilli() > account.ExpiresAt
}
